### DISCUSSION TITLE
for i, v in pairs on mouse down

### DISCUSSION THREAD

**Discussion by DRBAUS** [RomanKubiak/ctrlr#436] [General]
Hi there.....I'm back at it again.

After watching a video on YT I now understand how for i, v in pairs works and I did manage to get a working script upon 
Modulator Value Change.

But...I would like to have this function activated with a Mouse Down but somehow this does not seem possible. 

Is that correct or has any of you had this working?

Thanks,

Robin.

**Comment by dnaldoog** [comment]
It should be entirely possible- can you post the code?

**Reply by DRBAUS** [reply]
this is the code (that I got from you for the other panel). I made some adaptation but it works as intended...but not on Mouse Down.
Possibly this has to do with those lines in the top part of the screen.

`if source == 4 then
  sName = L(mod:getName())

        local t = {"TIME1", "TIME2"} 

        for i, v in pairs(t) do
            if sName == v then

			panel:getModulatorByName(v):getComponent():setProperty("componentLabelColour", "FFA20A0A", false)
			panel:getModulatorByName(v):getComponent():setProperty("componentVisibleName", "EDIT", false)
		
            end 
        end 
    end 
end`

**Reply by dnaldoog** [reply]
I wrote solution that actually wasn't working, so deleted it - this is much simpler. Have I got the logic right?

I'm not sure what your intention is, because once your function above is fired off the buttons (if that's what they are) will never change back again - so I added  default values at the beginning of the callback function.

<pre>
myMouseDown = function(--[[ CtrlrComponent --]] comp --[[ MouseEvent --]], event)
    panel:getModulatorByName("TIME1"):getComponent():setProperty("componentLabelColour", "ff000000", false)
    panel:getModulatorByName("TIME1"):getComponent():setProperty("componentVisibleName", "Time1", false)
    panel:getModulatorByName("TIME2"):getComponent():setProperty("componentLabelColour", "ff000000", false)
    panel:getModulatorByName("TIME2"):getComponent():setProperty("componentVisibleName", "Time2", false)

    sName = L(comp:getOwner():getName())
    panel:getModulatorByName(sName):getComponent():setProperty("componentLabelColour", "FFA20A0A", false)
    panel:getModulatorByName(sName):getComponent():setProperty("componentVisibleName", "EDIT", false)
end
</pre>

Also, you cannot use `source` or `L(mod:getName())` in a mousedown function. :)
[mousedown TIME1_2_0_a_2022-02-23_00-47.zip](https://github.com/RomanKubiak/ctrlr/files/8118062/mousedown.TIME1_2_0_a_2022-02-23_00-47.zip)

----

As per the comment below:

_substitute_:

<pre>
    comp:setProperty("componentLabelColour", "FFA20A0A", false)
    comp:setProperty("componentVisibleName", "EDIT", false)
</pre>

_for_

<pre>
    sName = L(comp:getOwner():getName())
    panel:getModulatorByName(sName):getComponent():setProperty("componentLabelColour", "FFA20A0A", false)
    panel:getModulatorByName(sName):getComponent():setProperty("componentVisibleName", "EDIT", false)
</pre>

**Reply by DRBAUS** [reply]
Great....that helps a lot!!!

In the first method I changed the mod:getName to comp:getOwner:getName and it works but as you say...they will always stay in that 'state'. 

The method using the default values makes more sense and saves me from writing 45 scripts where i can use 1 and that's what I was searching for.

So what I am doing in the panel is this:

1 Mouse down: puts a modulator in EDIT mode so to speak.
2 SET Value: slider AND a valuebox to set the value (they will be remote controlling each other as well)
3 Double Click will set the Modulator back to it's non-edit state and will send the current Value by MySysex

**Reply by dnaldoog** [reply]
> this is the code (that I got from you for the other panel). I made some adaptation but it works as intended...but not on Mouse Down. Possibly this has to do with those lines in the top part of the screen.
> 
> if source == 4 then sName = L(mod:getName())
> 
>
>     local t = {"TIME1", "TIME2"} 
> 
>     for i, v in pairs(t) do
>         if sName == v then
> 
> 		panel:getModulatorByName(v):getComponent():setProperty("componentLabelColour", "FFA20A0A", false)
> 		panel:getModulatorByName(v):getComponent():setProperty("componentVisibleName", "EDIT", false)
> 	
>         end 
>     end 
> end 
> 
> 
> end

Iterating through `t{}` in your code would need  `for i, v in ipairs(t) do` not `for i, v in pairs(t) do` because your table `t{}` contains indexed elements.
for example:
<pre>
local t = {"TIME1", "TIME2"} 
console(t[2])
</pre>
would print TIME2

For `pairs` you need a hashed table of key/value pairs:

<pre>
tk={TIME1="TIME 1",TIME2="TIME 2",["TIME 3"]="TIME 3"}
for k,v in pairs(tk) do
console(k.." "..v)
end

--TIME 3 TIME 3
--TIME1 TIME 1
--TIME2 TIME 2 note the order is not the same as the order in the table

  console(tk.TIME2) -- prints TIME 2
</pre>

**Comment by dnaldoog** [comment]
Hi @DRBAUS 

Here's a refinement of the panel for more than two buttons - You said you want the same functionality for 45 buttons? This code will scale up nicely. All you need to do is add the button names to the table `tab_all` - Note that in the mousedown function all you need is `comp:setProperty ...` **so `sName = L(comp:getOwner():getName())`  is not necessary.**

<pre>
--
-- Called when a mouse is down on this component
--

myMouseDown = function(--[[ CtrlrComponent --]] comp --[[ MouseEvent --]], event)
for _,v in ipairs(tab_all) do
reset(v)
end
    comp:setProperty("componentLabelColour", "FFA20A0A", false)
    comp:setProperty("componentVisibleName", "EDIT", false)
end
 

-- this loop uses ipairs, not pairs to run through each modulator and reset it
-- it would be better to set up lua variables in the init script to make this much faster
-- ie _G[control]:getComponent():setProperty("componentLabelColour", "ff000000", false)
--see panel for implementation of this

function reset(control)

	local Control=firstToUpper(control)

   -- panel:getModulatorByName(control):getComponent():setProperty("componentLabelColour", "ff000000", false)
   -- panel:getModulatorByName(control):getComponent():setProperty("componentVisibleName", Control, false)
    _G[control]:getComponent():setProperty("componentLabelColour", "ff000000", false)
    _G[control]:getComponent():setProperty("componentVisibleName", Control, false)

end

-- global table of all modulator names that will be passed to function
tab_all={"TIME1","TIME2","TIME3","TIME4","TIME5","TIME6"}

-- renames the label from TIME3 (the modulator name) to Time3
function firstToUpper(str)
	s=string.lower(str)
    return (s:gsub("^%l", string.upper))
end
</pre>

----

[mousedown TIME1_3_0_a_2022-02-23_07-45.zip](https://github.com/RomanKubiak/ctrlr/files/8120645/mousedown.TIME1_3_0_a_2022-02-23_07-45.zip)
